src/main/java/net/minecraftforge/event/AddValidBlocksToBlockEntityEvent.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.event;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.EntityBlock;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.minecraftforge.eventbus.api.Event;
+import net.minecraftforge.fml.event.IModBusEvent;
+import net.minecraftforge.registries.ForgeRegistries;
+import net.minecraftforge.registries.RegistryObject;
+
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Provided to add new blocks to a {@link BlockEntityType}'s list of valid blocks, used for ticking, rendering, and other purposes.
+ * For your own mod's blocks, specify the blocks in the {@link BlockEntityType.Builder}, don't use this event.
+ * Blocks added to the valid blocks list are expected to have attached the same {@link BlockEntity} as the parent type in their {@link EntityBlock} implementation.
+ *
+ * Fired on the Mod bus {@link IModBusEvent}.
+ */
+public class AddValidBlocksToBlockEntityEvent extends Event implements IModBusEvent
+{
+    private final Map<BlockEntityType<?>, Set<ResourceKey<Block>>> newBlocks;
+
+    @ApiStatus.Internal
+    public AddValidBlocksToBlockEntityEvent(Map<BlockEntityType<?>, Set<ResourceKey<Block>>> newBlocks)
+    {
+        this.newBlocks = newBlocks;
+    }
+
+    /**
+     * Adds a new {@code block} to the valid blocks list for the given {@code type}. It can be used with {@link RegistryObject}
+     */
+    public void addValidBlock(BlockEntityType<?> type, Supplier<? extends Block> supplier)
+    {
+        final Optional<ResourceKey<Block>> resourceKey = ForgeRegistries.BLOCKS.getResourceKey(supplier.get());
+        if (resourceKey.isEmpty())
+        {
+            throw new NullPointerException("Tried to add a not registered Block to BlockEntityType: " + ForgeRegistries.BLOCK_ENTITY_TYPES.getKey(type) + " Block Class: " + supplier.get().getClass().getSimpleName());
+        }
+        if (!newBlocks.containsKey(type))
+        {
+            newBlocks.put(type, new HashSet<>());
+        }
+        newBlocks.get(type).add(resourceKey.get());
```

Comment:
From what I understand, most "register"-style events throw if the entry has been registered multiple times; is it worth doing that here?

Reply:
I think you misunderstand, multiple mods can register to the same block entity type, this function just ensures that a set exists to add to when one is added for the first time.

Reply:
I'm talking about the `add(resourceKey.get())` - surely each block should only have a resource key added once? Would multiple mods be adding the same block to the block entity?

Reply:
Also, in that case, wouldn't it be cleaner to do `newBlocks.computeIfAbsent(type, x -> new HashSet<>()).add(resourceKey.get())` rather than a separate contains/put?

Reply:
You are reviewing an old version of the code. But yes it could use computeIfAbsent